File: lab3-primit/haskell/pltypeBool.hs
module Main where

import System.IO (hSetBuffering, stdout, BufferMode(..))
import Data.Char (toUpper)

main =
    do
    initialiseIO
    putStrLn $ "known PLs = " ++ show allPLs
    pl <- getElement "programming language"
    putStrLn $ show pl ++ " is of type: " ++ show (pl2PLType pl)
    
initialiseIO =
    do
    hSetBuffering stdout NoBuffering 
        -- ensure any console output is shown asap

data PL = JAVA | C | ADA | PYTHON | LISP | HASKELL | PROLOG
    deriving (Show, -- default formatting
              Read, -- default parsing
              Eq,   -- default equality testing
              Bounded, -- default minBound and maxBound
              Enum) -- default sequencing (needed for ranges)
data PLType = OO | PROCEDURAL | FUNCTIONAL | LOGICAL
    deriving (Show, -- default formatting
              Read, -- default parsing
              Eq,   -- default equality testing
              Bounded, -- default minBound and maxBound
              Enum) -- default sequencing (needed for ranges)

allPLs :: [PL] -- ie it is a list of PL elements
allPLs = [minBound .. maxBound]

pl2PLType pl 
    | (pl == JAVA || pl == PYTHON) = OO
    | (pl == C || pl == ADA) = PROCEDURAL
    | (pl == PROLOG) = LOGICAL
    | otherwise = FUNCTIONAL

{-
  `getElement'
  queries the user for one element until the user types something 
  that can be interpreted as the correct type of element (eg integer)
-}
getElement elementTypeName =
    do
    hSetBuffering stdout NoBuffering -- print to console in real time, not in batches
    putStr ("Input one " ++ elementTypeName ++ ": ") -- print the prompt
    line <- getLine -- get whatever user types as a string
    case parseElement line of
         Just element -> 
            do
            return element
         Nothing -> 
            do
            putStrLn ("Invalid " ++ elementTypeName ++ ", try again")
            getElement elementTypeName -- try again - using recursion

parseElement line =
    case reads line of
        [] -> -- no valid interpretation of the line as an element ([] = the empty list)
            Nothing
        -- [pattern 2:]
        ((e,_) : _) -> -- found at least one interpretation, call it "n"
            Just e -- type of e is derived from context
                   -- where getElement is used

File: lab3-primit/java/PLTypeBool.java
import java.util.Scanner;
import java.util.InputMismatchException;
import java.util.EnumSet;

public class PLTypeBool
{
    enum PL { JAVA, C, ADA, PYTHON, LISP, HASKELL, PROLOG }
    
    enum PLType { OO, PROCEDURAL, FUNCTIONAL, LOGICAL }
    
    private static <E extends Enum<E>> E getEnumElement(String elementTypeName, Class<E> elementType)
    {
        boolean haveResult = false;
        E result = null;
        Scanner stdin = new Scanner(System.in);
        
        while ( ! haveResult )
        {
            System.out.print("Input " + elementTypeName + ": ");
            try
            {
                result = Enum.valueOf(elementType, stdin.next().toUpperCase());
                haveResult = true;
            }
            catch (IllegalArgumentException e)
            {
                System.out.println("Not a valid " + elementTypeName + ".");
                stdin.nextLine(); // skip the invalid input
            }
        }
        
        return result;
    }
  
    private static PLType pl2PLType(PL pl)
    {
        PLType type = 
            (pl == PL.JAVA || pl == PL.PYTHON ? PLType.OO :
                (pl == PL.C || pl == PL.ADA ? PLType.PROCEDURAL :
                    (pl == PL.PROLOG ? PLType.LOGICAL : 
                        true ? PLType.FUNCTIONAL
                    )
                )
            );
        /* 
            !!!! Beware: The above is not a recommended programming style.
            !!!!         It is used here only to develop understanding of
            !!!!         functional and imperative if-then-else statements.
        */
        return type;
    }

    public static void main(String[] args)
    {
        System.out.print("Known PLs = ");
        for (PL t : EnumSet.allOf(PL.class)) 
        {
            System.out.print(t + " ");
        }
        System.out.println();
        
        PL pl = getEnumElement("programming language", PL.class);
        System.out.println(pl + " is of type: " + pl2PLType(pl));
    }
}